crates/ruff_server/src/server.rs

```diff
@@ -160,18 +114,69 @@ impl Server {
                     api::request(req)
                 }
                 lsp::Message::Notification(notification) => api::notification(notification),
-                lsp::Message::Response(response) => {
-                    tracing::error!(
-                        "Expected request or notification, got response instead: {response:?}"
-                    );
-                    continue;
-                }
+                lsp::Message::Response(response) => scheduler.response(response),
             };
             scheduler.dispatch(task);
         }
         Ok(())
     }
 
+    fn try_register_capabilities(
+        client_capabilities: &ClientCapabilities,
+        scheduler: &mut Scheduler,
+    ) {
+        let dynamic_registration = client_capabilities
+            .workspace
+            .as_ref()
+            .and_then(|workspace| workspace.did_change_watched_files)
+            .and_then(|watched_files| watched_files.dynamic_registration)
+            .unwrap_or_default();
+        if dynamic_registration {
+            // Register all dynamic capabilities here
+
+            // `workspace/didChangeWatchedFiles`
+            // (this registers the configuration file watcher)
+            let params = lsp_types::RegistrationParams {
+                registrations: vec![lsp_types::Registration {
+                    id: "ruff-server-watch".into(),
+                    method: "workspace/didChangeWatchedFiles".into(),
+                    register_options: Some(
+                        serde_json::to_value(DidChangeWatchedFilesRegistrationOptions {
+                            watchers: vec![
+                                FileSystemWatcher {
+                                    glob_pattern: types::GlobPattern::String(
+                                        "**/.?ruff.toml".into(),
+                                    ),
+                                    kind: None,
+                                },
+                                FileSystemWatcher {
+                                    glob_pattern: types::GlobPattern::String(
+                                        "**/pyproject.toml".into(),
+                                    ),
+                                    kind: None,
+                                },
+                            ],
+                        })
+                        .unwrap(),
+                    ),
+                }],
+            };
+
+            let response_handler = |()| {
+                tracing::info!("Configuration file watcher successfully registered");
+                Task::nothing()
+            };
+
+            if let Err(err) = scheduler
+                .request::<lsp_types::request::RegisterCapability>(params, response_handler)
+            {
+                tracing::error!("An error occurred when trying to register the configuration file watcher: {err}");
+            }
```

Comment:
Nit: This is not something we need to immediately follow up. It's more an observation and maybe we have ideas on how we could improve it later on. 

What would be nice is if it wouldn't be necessary to specify the `Request` type when calling the function, and instead could let Rust infer it. I think the issue here is that we pass the parameters but there's no way to get to the request from the parameters (only the other way around). 
I don't have a solution to it, but thought it might be worth raising because maybe you have a great idea on how this could be improved.

Reply:
Unfortunately I think this is just a limitation of Rust's type resolver, and I understand why. It would definitely be nice if the compiler could infer the type from the `Params` + `Result` type, but it would fail in a scenario where multiple requests had the same `Params` and `Result` type.

I'm fine with being explicit about the request type, to be honest. We do something similar when creating tasks for API endpoints. We *could* make the request a function argument and that would remove the need to specify the generic parameter, but I prefer putting it in angle brackets 😄